common/malloc_utils/src/glibc.rs

```diff
@@ -0,0 +1,194 @@
+    );
+}
+
+/// Calls `mallinfo` and updates Prometheus metrics with the results.
+pub fn scrape_mallinfo_metrics() {
+    // The docs for this function say it is thread-unsafe since it may return inconsistent results.
+    // Since these are just metrics it's not a concern to us if they're sometimes inconsistent.
+    //
+    // Docs:
+    //
+    // https://man7.org/linux/man-pages/man3/mallinfo.3.html
+    let mallinfo = unsafe { libc::mallinfo() };
+
+    set_gauge(&MALLINFO_ARENA, mallinfo.arena as i64);
+    set_gauge(&MALLINFO_ORDBLKS, mallinfo.ordblks as i64);
+    set_gauge(&MALLINFO_SMBLKS, mallinfo.smblks as i64);
+    set_gauge(&MALLINFO_HBLKS, mallinfo.hblks as i64);
+    set_gauge(&MALLINFO_HBLKHD, mallinfo.hblkhd as i64);
+    set_gauge(&MALLINFO_FSMBLKS, mallinfo.fsmblks as i64);
+    set_gauge(&MALLINFO_UORDBLKS, mallinfo.uordblks as i64);
+    set_gauge(&MALLINFO_FORDBLKS, mallinfo.fordblks as i64);
+    set_gauge(&MALLINFO_KEEPCOST, mallinfo.keepcost as i64);
+}
+
+/// Perform all configuration routines.
+pub fn configure_glibc_malloc() -> Result<(), String> {
+    if !env_var_present(ENV_VAR_ARENA_MAX) {
+        let arena_max = match OPTIMAL_ARENA_MAX {
+            ArenaMaxSetting::DoNotSet => None,
+            ArenaMaxSetting::Fixed(n) => Some(n),
+            ArenaMaxSetting::NumCpus => Some(num_cpus::get() as c_int),
+        };
+
+        if let Some(max) = arena_max {
+            if let Err(e) = malloc_arena_max(max) {
+                return Err(format!("failed (code {}) to set malloc max arena count", e));
+            }
+        }
+    }
+
+    if !env_var_present(ENV_VAR_MMAP_THRESHOLD) {
+        if let Err(e) = malloc_mmap_threshold(OPTIMAL_MMAP_THRESHOLD) {
+            return Err(format!("failed (code {}) to set malloc mmap threshold", e));
+        }
+    }
+
+    Ok(())
+}
+
+/// Returns `true` if an environment variable is present.
+fn env_var_present(name: &str) -> bool {
+    env::var(name) != Err(env::VarError::NotPresent)
+}
+
+/// Uses `mallopt` to set the `M_ARENA_MAX` value, specifying the number of memory arenas to be
+/// created by malloc.
+///
+/// Generally speaking, a smaller arena count reduces memory fragmentation at the cost of memory contention
+/// between threads.
+///
+/// ## Resources
+///
+/// - https://man7.org/linux/man-pages/man3/mallopt.3.html
+fn malloc_arena_max(num_arenas: c_int) -> Result<(), c_int> {
+    unsafe { into_result(ffi::mallopt(M_ARENA_MAX, num_arenas)) }
+}
+
+/// Uses `mallopt` to set the `M_MMAP_THRESHOLD` value, specifying the threshold where objects of this
+/// size or larger are allocated via an `mmap`.
+///
+/// ## Resources
+///
+/// - https://man7.org/linux/man-pages/man3/mallopt.3.html
+fn malloc_mmap_threshold(num_arenas: c_int) -> Result<(), c_int> {
+    unsafe { into_result(ffi::mallopt(M_MMAP_THRESHOLD, num_arenas)) }
+}
+
+mod ffi {
+    extern "C" {
+        pub fn mallopt(
```

Comment:
Is there anything preventing us from using `libc`'s `mallopt`? https://docs.rs/libc/0.2.95/libc/fn.mallopt.html

Reply:
I can't recall why this is the case. I'll try switching to the `libc` version.